pkg/gadgets/top/ebpf/tracer/tracer.go

```diff
@@ -0,0 +1,249 @@
+// Copyright 2019-2022 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracer
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/kinvolk/inspektor-gadget/pkg/bpfstats"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/top/ebpf/piditer"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgets/top/ebpf/types"
+
+	"github.com/cilium/ebpf"
+)
+
+type Config struct {
+	MaxRows  int
+	Interval time.Duration
+	SortBy   types.SortBy
+}
+
+type Tracer struct {
+	config        *Config
+	statsCallback func([]types.Stats)
+	errorCallback func(error)
+	done          chan bool
+
+	iter                *piditer.PidIter
+	useFallbackIterator bool
+}
+
+func NewTracer(config *Config, statsCallback func([]types.Stats), errorCallback func(error),
+) (*Tracer, error) {
+	t := &Tracer{
+		config:        config,
+		statsCallback: statsCallback,
+		errorCallback: errorCallback,
+		done:          make(chan bool),
+	}
+
+	if err := t.start(); err != nil {
+		t.Stop()
+		return nil, err
+	}
+
+	t.run()
+
+	return t, nil
+}
+
+func (t *Tracer) start() error {
+	// Enable stats collection
+	err := bpfstats.EnableBPFStats()
+	if err != nil {
+		return err
+	}
+
+	t.useFallbackIterator = false
+
+	// To resolve pids, we will first try to iterate using a bpf
+	// program. If that doesn't work, we will fall back to scanning
+	// all used fds in all processes /proc/$pid/fdinfo/$fd.
+	iter, err := piditer.NewTracer()
+	if err != nil {
+		t.useFallbackIterator = true
+	} else {
+		t.iter = iter
+	}
+
+	return nil
+}
+
+func (t *Tracer) Stop() {
+	close(t.done)
+
+	if t.iter != nil {
+		t.iter.Close()
+	}
+
+	bpfstats.DisableBPFStats()
+}
+
+func getPidMapFromPids(pids []*piditer.PidIterEntry) map[uint32][]*types.PidInfo {
+	pidmap := make(map[uint32][]*types.PidInfo)
+	for _, e := range pids {
+		if _, ok := pidmap[e.ProgID]; !ok {
+			pidmap[e.ProgID] = make([]*types.PidInfo, 0, 1)
+		}
+		pidmap[e.ProgID] = append(pidmap[e.ProgID], &types.PidInfo{
+			Pid:  e.Pid,
+			Comm: e.Comm,
+		})
+	}
+	return pidmap
+}
+
+func getPidMapFromProcFs() map[uint32][]*types.PidInfo {
+	pidmap := make(map[uint32][]*types.PidInfo)
+	processes, err := os.ReadDir("/proc/")
+	if err == nil {
```

Comment:
Agree with the early return :)

About the error: as we're "just" enriching data here, I'm not too sure about what we should do when there's an error. If we're still getting all the other info, it might be worth to show it without the pid info. Maybe we should do the same with the `t.iter.DumpPids()` and ignore that error as well and try the fallback solution in case of an error.

Reply:
I agree that the information produced in this step is "extra" information and not fundamental. My concern is how do we indicate that we failed to enrich the output, i.e., how can a user distinguish the case when no pids are using a program vs the case when we failed to enrich the data?  (I don't want to produce any misleading information). Maybe it's easier to fail in this case to avoid ambiguity?

Reply:
I see. I'd propose to fail for the time being, then we can implement the mechanism you described. (I'd like to get this PR in, I think it's almost complete and we can make other iterations later on to improve it, otherwise it becomes too complicated to review it).

Reply:
One option could be to make `getPidMapFromProcFs` a `Tracer`'s method so it can call `errorCallback`. Of course, this proposal can be explored in another PR.

Reply:
We can discuss further in #794.